hip_data_tools/google/adwords.py

```diff
@@ -0,0 +1,115 @@
+        developer_token_var (str): The Key pointer for developer_token
+    """
+
+    def __init__(self,
+                 source: KeyValueSource = ENVIRONMENT,
+                 client_secret_var: str = "adwords_client_secret",
+                 refresh_token_var: str = "adwords_refresh_token",
+                 developer_token_var="adwords_developer_token"):
+        self._required_keys = [client_secret_var, refresh_token_var, developer_token_var, ]
+        super().__init__(self._required_keys, source)
+        self.client_secret = self.get_secret(client_secret_var)
+        self.refresh_token = self.get_secret(refresh_token_var)
+        self.developer_key = self.get_secret(developer_token_var)
+
+
+@dataclass
+class GoogleAdWordsConnectionSettings:
+    """
+    Handle connection settings for AdWords client
+    """
+    client_id: str
+    user_agent: str
+    client_customer_id: str
+    secrets_manager: GoogleAdWordsSecretsManager
+
+
+class GoogleAdWordsConnectionManager:
+    """
+    AdWords Connection Manager that manages the lifecycle of authentication and the resulting
+    adwords client
+    Args:
+        settings (GoogleAdWordsConnectionSettings): setting to use for connecting to adwords
+    """
+
+    def __init__(self, settings: GoogleAdWordsConnectionSettings):
+        self.settings = settings
+        self._adwords_client = None
+
+    def get_adwords_client(self) -> AdWordsClient:
+        """
+        Gets the connected  adwords client, and creates one if not exist
+        Returns:
+
+        """
+        if self._adwords_client is None:
+            self._adwords_client = self._create_adwords_client()
+        return self._adwords_client
+
+    def _create_adwords_client(self) -> AdWordsClient:
+        oauth2_client = self._get_oauth_client()
+        adwords_client = AdWordsClient(
+            developer_token=self.settings.secrets_manager.developer_key,
+            oauth2_client=oauth2_client,
+            user_agent=self.settings.user_agent)
+        return adwords_client
+
+    def _get_oauth_client(self) -> GoogleOAuth2Client:
+        oauth2_client = oauth2.GoogleRefreshTokenClient(
+            client_id=self.settings.client_id,
+            client_secret=self.settings.secrets_manager.client_secret,
+            refresh_token=self.settings.secrets_manager.refresh_token
+        )
+        return oauth2_client
+
+
+class AdWordsUtil:
+    """
+    Generic Adwords Utility that generates the required services that perform actions on adwords
+    Args:
+        conn (GoogleAdWordsConnectionManager): Connection manager to handle the creation of
+        adwords client
+    """
+
+    def __init__(self, conn: GoogleAdWordsConnectionManager):
+        self.conn = conn
+
+    def _get_customer_service(self):
+        return self.conn.get_adwords_client().GetService('CustomerService', version='v201809')
+
+
+class AdWordsCustomerUtil(AdWordsUtil):
+    """
+    Adwords Utility to handle customer details
```

Comment:
what do you mean by Customers here? I'm don't think i'm up with that lingo

Reply:
Customers as per Google Adwords Client.

Reply:
Customers of Google Adwords

Reply:
right, can you be more specific? That's quite a generic term.. do you mean API clients? People who buy a job from us? Chris Hobbs. The guy who runs My Mum Made Me Paint My house (genuine tradie. look him up)? 

I suppose what I'm saying is that's quite a generic term and is a bit confusing.

Reply:
Adwords is confusing

Reply:
Customers of AdWords... people who pay money to Google, and then create accounts in Google AdWords, so their businesses can get traffic. Or maybe just because they want to waste money...
I don't know what makes humans tick.